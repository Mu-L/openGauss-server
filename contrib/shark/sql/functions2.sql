create schema functions_test2;
set search_path = 'functions_test2';

select datediff(year, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(yy, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(yyyy, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(quarter, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(qq, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(q, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(month, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(mm, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(m, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(dayofyear, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(dy, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(y, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(day, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(dd, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(d, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(week, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(wk, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(ww, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(weekday, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(dw, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(w, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(hour, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(hh, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(minute,CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(mi, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(n, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(second, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(ss, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(s, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(millisecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
select datediff(ms, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
select datediff(microsecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
select datediff(mcs, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
select datediff(nanosecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
select datediff(ns, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));

select datediff_big(year, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(quarter, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(dayofyear, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(day, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(week,CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(hour, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(minute,CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(second, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(millisecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
select datediff_big(microsecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
select datediff_big(nanosecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));

select datediff(month, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
select datediff(month, CAST('2037-03-01 23:30:05.523'AS timestamptz), CAST('2036-02-28 23:30:05.523'AS timestamptz));
select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS timestamptz), CAST('2036-02-28 23:30:05.523'AS timestamptz));
select datediff(month, CAST('2037-03-01 23:30:05.523'AS date), CAST('2036-02-28 23:30:05.523'AS date));
select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS date), CAST('2036-02-28 23:30:05.523'AS date));
select datediff(month, CAST('2037-03-01 23:30:05.523'AS time), CAST('2036-02-28 23:30:05.523'AS time));
select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS time), CAST('2036-02-28 23:30:05.523'AS time));
select datediff(month, CAST('2037-03-01 23:30:05.523'AS smalldatetime), CAST('2036-02-28 23:30:05.523'AS smalldatetime));
select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS smalldatetime), CAST('2036-02-28 23:30:05.523'AS smalldatetime));

select datediff(seconds, CAST(make_timestamp(1, 2, 12, 0, 0, 0) AS date), CAST(make_timestamp(10000, 2, 12, 0, 0, 0) AS date));
select datediff_big(nanoseconds, CAST(make_timestamp(1, 2, 12, 0, 0, 0) AS date), CAST(make_timestamp(10000, 2, 12, 0, 0, 0) AS date));

select datediff(abc, CAST(make_timestamp(1, 2, 12, 0, 0, 0) AS date), CAST(make_timestamp(2000, 2, 12, 0, 0, 0) AS date));
select datediff(years, CAST(make_timestamp(1, 2, 12, 0, 0, 0) AS date), CAST(make_timestamp(2000, 2, 12, 0, 0, 0) AS date));

select CHARINDEX('hello', 'hello world');
select CHARINDEX('hello  ', 'hello world');
select CHARINDEX('hello world', 'hello');
select CHARINDEX(NULL, NULL);
select CHARINDEX(NULL, 'string');
select CHARINDEX('pattern', NULL);
select CHARINDEX('pattern', 'string', NULL);
select CHARINDEX('hello', 'hello world', -1);
select CHARINDEX('hello', 'hello world', 0);
select CHARINDEX('hello', 'hello world', 1);
select CHARINDEX('hello', 'hello world', 2);
select CHARINDEX('world', 'hello world', 6);
select CHARINDEX('world', 'hello world', 7);
select CHARINDEX('world', 'hello world', 8);
select CHARINDEX('is', 'This is a string');
select CHARINDEX('is', 'This is a string', 4);

select atn2(1.23423::int1, 2.3412::int1);
select atn2(1.23423::int2, 2.3412::int2);
select atn2(1.23423::int4, 2.3412::int4);
select atn2(1.23423::int8, 2.3412::int8);
select atn2(1.23423::real, 2.3412::real);
select atn2(1.23423::numeric, 2.3412::numeric);
select atn2('1.23423', '2.3412');

select atn2(2::int4, 1.2345::real);
select atn2(1.2345::double precision, 1234::int4);
select atn2('123,345', 123);

CREATE VIEW atn2_1 AS (
    SELECT
        ATN2(CAST(2 AS INT), CAST(3 AS INT)) AS res1,
        ATN2(CAST(2.5 AS FLOAT), CAST(3.5 AS FLOAT)) AS res2,
        ATN2(CAST(2.5 AS REAL), CAST(3.5 AS REAL)) AS res3,
        ATN2(CAST(2.5 AS BIGINT), CAST(3.5 AS BIGINT)) AS res4,
        ATN2(CAST(2.5 AS SMALLINT), CAST(3.5 AS SMALLINT)) AS res5,
        ATN2(CAST(2.5 AS TINYINT), CAST(3.5 AS TINYINT)) AS res6,
        ATN2(CAST(2.5 AS DECIMAL), CAST(3.5 AS DECIMAL)) AS res7,
        ATN2(CAST(2.5 AS NUMERIC), CAST(3.5 AS NUMERIC)) AS res8,
        ATN2(CAST('2.5' AS CHAR), CAST('3.5' AS CHAR)) AS res9,
        ATN2(CAST('2.5' AS VARCHAR), CAST('3.5' AS VARCHAR)) AS res10,
        ATN2(CAST('2.5' AS BPCHAR), CAST('3.5' AS BPCHAR)) AS res11,
        ATN2(CAST('2.5' AS NVARCHAR2), CAST('3.5' AS NVARCHAR2)) AS res12
    );
select * from atn2_1;
drop view atn2_1;


select ISNULL(NULL, 1);
select ISNULL(2, 1);
select ISNULL();
select ISNULL();

select log10(10);
select log10(10::real);
select log10(10::double precision);
select log10(10::numeric);

select atn2('abc', '2.3412');
select atn2(123::varbinary, '2343'::varbinary);

select log10('abc');
select log10(10::vharchar);
select log10(now());

select ISNULL('abc', 123);

reset search_path;
drop schema functions_test2 cascade;
