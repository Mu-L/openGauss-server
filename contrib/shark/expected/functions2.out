create schema functions_test2;
set search_path = 'functions_test2';
select datediff(year, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
       -1
(1 row)

select datediff(yy, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
       -1
(1 row)

select datediff(yyyy, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
       -1
(1 row)

select datediff(quarter, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
       -4
(1 row)

select datediff(qq, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
       -4
(1 row)

select datediff(q, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
       -4
(1 row)

select datediff(month, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
      -13
(1 row)

select datediff(mm, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
      -13
(1 row)

select datediff(m, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
      -13
(1 row)

select datediff(dayofyear, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(dy, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(y, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(day, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(dd, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(d, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(week, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
      -52
(1 row)

select datediff(wk, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
      -52
(1 row)

select datediff(ww, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
      -52
(1 row)

select datediff(weekday, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(dw, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(w, CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
     -367
(1 row)

select datediff(hour, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
    -8808
(1 row)

select datediff(hh, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
    -8808
(1 row)

select datediff(minute,CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
  -528480
(1 row)

select datediff(mi, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
  -528480
(1 row)

select datediff(n, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
  -528480
(1 row)

select datediff(second, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff  
-----------
 -31708800
(1 row)

select datediff(ss, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff  
-----------
 -31708800
(1 row)

select datediff(s, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff  
-----------
 -31708800
(1 row)

select datediff(millisecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
 datediff 
----------
     -111
(1 row)

select datediff(ms, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
 datediff 
----------
     -111
(1 row)

select datediff(microsecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
 datediff 
----------
  -111000
(1 row)

select datediff(mcs, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
 datediff 
----------
  -111000
(1 row)

select datediff(nanosecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
  datediff  
------------
 -111000000
(1 row)

select datediff(ns, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
  datediff  
------------
 -111000000
(1 row)

select datediff_big(year, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
           -1
(1 row)

select datediff_big(quarter, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
           -4
(1 row)

select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
          -13
(1 row)

select datediff_big(dayofyear, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
         -367
(1 row)

select datediff_big(day, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
         -367
(1 row)

select datediff_big(week,CAST('2037-03-01 23:30:05.523'AS timestamp),CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
          -52
(1 row)

select datediff_big(hour, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
        -8808
(1 row)

select datediff_big(minute,CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
      -528480
(1 row)

select datediff_big(second, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
    -31708800
(1 row)

select datediff_big(millisecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
 datediff_big 
--------------
         -111
(1 row)

select datediff_big(microsecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
 datediff_big 
--------------
      -111000
(1 row)

select datediff_big(nanosecond, CAST('2036-02-28 01:23:45.234'AS timestamp), CAST('2036-02-28 01:23:45.123'AS timestamp));
 datediff_big 
--------------
   -111000000
(1 row)

select datediff(month, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff 
----------
      -13
(1 row)

select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS timestamp), CAST('2036-02-28 23:30:05.523'AS timestamp));
 datediff_big 
--------------
          -13
(1 row)

select datediff(month, CAST('2037-03-01 23:30:05.523'AS timestamptz), CAST('2036-02-28 23:30:05.523'AS timestamptz));
 datediff 
----------
      -13
(1 row)

select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS timestamptz), CAST('2036-02-28 23:30:05.523'AS timestamptz));
 datediff_big 
--------------
          -13
(1 row)

select datediff(month, CAST('2037-03-01 23:30:05.523'AS date), CAST('2036-02-28 23:30:05.523'AS date));
 datediff 
----------
      -13
(1 row)

select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS date), CAST('2036-02-28 23:30:05.523'AS date));
 datediff_big 
--------------
          -13
(1 row)

select datediff(month, CAST('2037-03-01 23:30:05.523'AS time), CAST('2036-02-28 23:30:05.523'AS time));
 datediff 
----------
        0
(1 row)

select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS time), CAST('2036-02-28 23:30:05.523'AS time));
 datediff_big 
--------------
            0
(1 row)

select datediff(month, CAST('2037-03-01 23:30:05.523'AS smalldatetime), CAST('2036-02-28 23:30:05.523'AS smalldatetime));
 datediff 
----------
      -13
(1 row)

select datediff_big(month, CAST('2037-03-01 23:30:05.523'AS smalldatetime), CAST('2036-02-28 23:30:05.523'AS smalldatetime));
 datediff_big 
--------------
          -13
(1 row)

select datediff(seconds, CAST(make_timestamp(1, 2, 12, 0, 0, 0) AS date), CAST(make_timestamp(10000, 2, 12, 0, 0, 0) AS date));
ERROR:  The datediff function resulted in an overflow. The number of dateparts separating two date/time instances is too large. Try to use datediff with a less precise datepart
CONTEXT:  PL/pgSQL function datediff(text,date,date) line 3 at RETURN
referenced column: datediff
select datediff_big(nanoseconds, CAST(make_timestamp(1, 2, 12, 0, 0, 0) AS date), CAST(make_timestamp(10000, 2, 12, 0, 0, 0) AS date));
ERROR:  'nanoseconds' is not a recognized datediff option
CONTEXT:  PL/pgSQL function datediff_big(text,date,date) line 3 at RETURN
referenced column: datediff_big
select datediff(abc, CAST(make_timestamp(1, 2, 12, 0, 0, 0) AS date), CAST(make_timestamp(2000, 2, 12, 0, 0, 0) AS date));
ERROR:  'abc' is not a recognized datediff option
CONTEXT:  PL/pgSQL function datediff(text,date,date) line 3 at RETURN
referenced column: datediff
select datediff(years, CAST(make_timestamp(1, 2, 12, 0, 0, 0) AS date), CAST(make_timestamp(2000, 2, 12, 0, 0, 0) AS date));
 datediff 
----------
     1999
(1 row)

select CHARINDEX('hello', 'hello world');
 charindex 
-----------
         1
(1 row)

select CHARINDEX('hello  ', 'hello world');
 charindex 
-----------
         0
(1 row)

select CHARINDEX('hello world', 'hello');
 charindex 
-----------
         0
(1 row)

select CHARINDEX(NULL, NULL);
 charindex 
-----------
          
(1 row)

select CHARINDEX(NULL, 'string');
 charindex 
-----------
          
(1 row)

select CHARINDEX('pattern', NULL);
 charindex 
-----------
          
(1 row)

select CHARINDEX('pattern', 'string', NULL);
 charindex 
-----------
          
(1 row)

select CHARINDEX('hello', 'hello world', -1);
 charindex 
-----------
         1
(1 row)

select CHARINDEX('hello', 'hello world', 0);
 charindex 
-----------
         1
(1 row)

select CHARINDEX('hello', 'hello world', 1);
 charindex 
-----------
         1
(1 row)

select CHARINDEX('hello', 'hello world', 2);
 charindex 
-----------
         0
(1 row)

select CHARINDEX('world', 'hello world', 6);
 charindex 
-----------
         7
(1 row)

select CHARINDEX('world', 'hello world', 7);
 charindex 
-----------
         7
(1 row)

select CHARINDEX('world', 'hello world', 8);
 charindex 
-----------
         0
(1 row)

select CHARINDEX('is', 'This is a string');
 charindex 
-----------
         3
(1 row)

select CHARINDEX('is', 'This is a string', 4);
 charindex 
-----------
         6
(1 row)

select atn2(1.23423::int1, 2.3412::int1);
       atn2       
------------------
 .463647609000806
(1 row)

select atn2(1.23423::int2, 2.3412::int2);
       atn2       
------------------
 .463647609000806
(1 row)

select atn2(1.23423::int4, 2.3412::int4);
       atn2       
------------------
 .463647609000806
(1 row)

select atn2(1.23423::int8, 2.3412::int8);
       atn2       
------------------
 .463647609000806
(1 row)

select atn2(1.23423::real, 2.3412::real);
       atn2       
------------------
 .485153152627375
(1 row)

select atn2(1.23423::numeric, 2.3412::numeric);
       atn2       
------------------
 .485153158718424
(1 row)

select atn2('1.23423', '2.3412');
       atn2       
------------------
 .485153158718424
(1 row)

select atn2(2::int4, 1.2345::real);
       atn2       
------------------
 1.01778944809654
(1 row)

select atn2(1.2345::double precision, 1234::int4);
        atn2        
--------------------
 .00100040485264725
(1 row)

select atn2('123,345', 123);
ERROR:  invalid input syntax for type double precision: "123,345"
LINE 1: select atn2('123,345', 123);
                    ^
CONTEXT:  referenced column: atn2
CREATE VIEW atn2_1 AS (
    SELECT
        ATN2(CAST(2 AS INT), CAST(3 AS INT)) AS res1,
        ATN2(CAST(2.5 AS FLOAT), CAST(3.5 AS FLOAT)) AS res2,
        ATN2(CAST(2.5 AS REAL), CAST(3.5 AS REAL)) AS res3,
        ATN2(CAST(2.5 AS BIGINT), CAST(3.5 AS BIGINT)) AS res4,
        ATN2(CAST(2.5 AS SMALLINT), CAST(3.5 AS SMALLINT)) AS res5,
        ATN2(CAST(2.5 AS TINYINT), CAST(3.5 AS TINYINT)) AS res6,
        ATN2(CAST(2.5 AS DECIMAL), CAST(3.5 AS DECIMAL)) AS res7,
        ATN2(CAST(2.5 AS NUMERIC), CAST(3.5 AS NUMERIC)) AS res8,
        ATN2(CAST('2.5' AS CHAR), CAST('3.5' AS CHAR)) AS res9,
        ATN2(CAST('2.5' AS VARCHAR), CAST('3.5' AS VARCHAR)) AS res10,
        ATN2(CAST('2.5' AS BPCHAR), CAST('3.5' AS BPCHAR)) AS res11,
        ATN2(CAST('2.5' AS NVARCHAR2), CAST('3.5' AS NVARCHAR2)) AS res12
    );
select * from atn2_1;
       res1       |       res2       |       res3       |       res4       |       res5       |       res6       |       res7       |       res8       |       res9       |      res10       |      res11       |      res12       
------------------+------------------+------------------+------------------+------------------+------------------+------------------+------------------+------------------+------------------+------------------+------------------
 .588002603547568 | .620249485982821 | .620249485982821 | .643501108793284 | .643501108793284 | .643501108793284 | .620249485982821 | .620249485982821 | .588002603547568 | .620249485982821 | .620249485982821 | .620249485982821
(1 row)

drop view atn2_1;
select ISNULL(NULL, 1);
 isnull 
--------
      1
(1 row)

select ISNULL(2, 1);
 isnull 
--------
      2
(1 row)

select ISNULL();
ERROR:  function isnull() does not exist
LINE 1: select ISNULL();
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
CONTEXT:  referenced column: isnull
select ISNULL();
ERROR:  function isnull() does not exist
LINE 1: select ISNULL();
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
CONTEXT:  referenced column: isnull
select log10(10);
 log10 
-------
     1
(1 row)

select log10(10::real);
 log10 
-------
     1
(1 row)

select log10(10::double precision);
 log10 
-------
     1
(1 row)

select log10(10::numeric);
 log10 
-------
     1
(1 row)

select atn2('abc', '2.3412');
ERROR:  invalid input syntax for type double precision: "abc"
CONTEXT:  PL/pgSQL function atn2(character varying,character varying) line 3 at RETURN
referenced column: atn2
select atn2(123::varbinary, '2343'::varbinary);
ERROR:  function atn2(varbinary, varbinary) does not exist
LINE 1: select atn2(123::varbinary, '2343'::varbinary);
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
CONTEXT:  referenced column: atn2
select log10('abc');
ERROR:  invalid input syntax for type double precision: "abc"
LINE 1: select log10('abc');
                     ^
CONTEXT:  referenced column: log10
select log10(10::vharchar);
ERROR:  type "vharchar" does not exist
LINE 1: select log10(10::vharchar);
                         ^
CONTEXT:  referenced column: log10
select log10(now());
ERROR:  function log10(timestamp with time zone) does not exist
LINE 1: select log10(now());
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
CONTEXT:  referenced column: log10
select ISNULL('abc', 123);
ERROR:  invalid input syntax for integer: "abc"
LINE 1: select ISNULL('abc', 123);
                      ^
CONTEXT:  referenced column: isnull
reset search_path;
drop schema functions_test2 cascade;
